x/reports/types/genesis_test.go

```diff
@@ -0,0 +1,210 @@
+package types_test
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/desmos-labs/desmos/v3/x/reports/types"
+)
+
+func TestValidateGenesis(t *testing.T) {
+	testCases := []struct {
+		name      string
+		data      *types.GenesisState
+		shouldErr bool
+	}{
+		{
+			name: "duplicated subspaces data returns error",
+			data: types.NewGenesisState([]types.SubspaceDataEntry{
+				types.NewSubspacesDataEntry(1, 1, 1),
+				types.NewSubspacesDataEntry(1, 1, 1),
+			}, nil, nil, types.DefaultParams()),
+			shouldErr: true,
+		},
+		{
+			name: "invalid subspaces data returns error",
+			data: types.NewGenesisState([]types.SubspaceDataEntry{
+				types.NewSubspacesDataEntry(0, 1, 1),
+			}, nil, nil, types.DefaultParams()),
+			shouldErr: true,
+		},
+		{
+			name: "duplicated reason returns error",
+			data: types.NewGenesisState(nil, []types.Reason{
+				types.NewReason(1, 1, "Spam", ""),
+				types.NewReason(1, 1, "Spam", ""),
+			}, nil, types.DefaultParams()),
+			shouldErr: true,
+		},
+		{
+			name: "invalid reason id returns error",
+			data: types.NewGenesisState(
+				[]types.SubspaceDataEntry{
+					types.NewSubspacesDataEntry(1, 1, 1),
+				},
+				[]types.Reason{
+					types.NewReason(1, 1, "Spam", ""),
+				},
+				nil,
+				types.DefaultParams(),
+			),
+			shouldErr: true,
+		},
+		{
+			name: "duplicated report returns error",
+			data: types.NewGenesisState(nil, nil, []types.Report{
+				types.NewReport(
+					1,
+					1,
+					1,
+					"",
+					types.NewPostTarget(1),
+					"cosmos1atdl3cpms89md5qa3rxtql0drtgftch2zgkr7v",
+					time.Date(2020, 1, 1, 12, 00, 00, 000, time.UTC),
+				),
+				types.NewReport(
+					1,
+					1,
+					1,
+					"",
+					types.NewPostTarget(1),
+					"cosmos1x85xq5m2ehkjzw928j9zfv3awdy0hqtnhrp9r6",
+					time.Date(2020, 1, 1, 12, 00, 00, 000, time.UTC),
+				),
+			}, types.DefaultParams()),
+			shouldErr: true,
+		},
+		{
+			name: "invalid report id returns error",
+			data: types.NewGenesisState(
+				[]types.SubspaceDataEntry{
+					types.NewSubspacesDataEntry(1, 1, 1),
+				},
+				nil,
+				[]types.Report{
+					types.NewReport(
+						1,
+						1,
+						1,
+						"",
+						types.NewPostTarget(1),
+						"cosmos1atdl3cpms89md5qa3rxtql0drtgftch2zgkr7v",
+						time.Date(2020, 1, 1, 12, 00, 00, 000, time.UTC),
+					),
+				},
+				types.DefaultParams(),
+			),
+			shouldErr: true,
+		},
+		{
+			name: "invalid report returns error",
+			data: types.NewGenesisState(
+				[]types.SubspaceDataEntry{
+					types.NewSubspacesDataEntry(1, 1, 2),
+				},
+				nil,
+				[]types.Report{
+					types.NewReport(
+						0,
```

Comment:
The subspace 0 has no next report id so the invalid report test is uncovered.
The test data should be:
```
types.NewReport(
    1,
    0,
    1,
    ...
)
```

Reply:
I've removed the reason id and report id checks from `ValidateGenesis` and added the reason id check inside the invariants. This should make it easier and safer to perform all the checks properly: during the genesis initialization all data will be stored but, as soon as invariants checks are performed, an error will be thrown halting the chain.

Reply:
The genesis height could be non-zero and the invariant checker is based on `InvCheckPeriod`, meaning that the genesis height might not be the invariant check point.  As a result, desmos cannot detect the failure within report_id or reason_id when importing genesis. Honestly, I don't think we should remove the id checks from `ValidateGenesis`.

Reply:
Actually invariants are always checked inside the `x/crisis` module `InitGenesis` method, unless you explicitly tell to *not* do so (see [here](https://github.com/cosmos/cosmos-sdk/blob/main/x/crisis/module.go#L143)). This is also why such module should always be the last one to be initialized

Reply:
Got it, thanks your explanation.